cosmwasm/enclaves/execute/src/registration/onchain.rs

```diff
@@ -7,19 +7,83 @@ use std::panic;
 use enclave_ffi_types::NodeAuthResult;
 
 use crate::registration::seed_exchange::SeedType;
+use crate::registration::attestation::verify_quote_ecdsa;
+use crate::registration::cert::verify_ra_report;
+
 use enclave_crypto::consts::OUTPUT_ENCRYPTED_SEED_SIZE;
 use enclave_crypto::PUBLIC_KEY_SIZE;
 use enclave_utils::{
     oom_handler::{self, get_then_clear_oom_happened},
     validate_const_ptr, validate_mut_ptr,
 };
 
+use sgx_types::sgx_ql_qv_result_t;
+
+#[cfg(feature = "SGX_MODE_HW")]
+use enclave_crypto::consts::SigningMethod;
+
 use super::cert::verify_ra_cert;
 use super::seed_exchange::encrypt_seed;
+use core::mem;
+use std::slice;
 
 #[cfg(feature = "light-client-validation")]
 use enclave_contract_engine::check_cert_in_current_block;
 
+#[cfg(feature = "light-client-validation")]
+use block_verifier::VERIFIED_BLOCK_MESSAGES;
+
+#[cfg(feature = "light-client-validation")]
+fn get_current_block_time_s() -> i64
+{
+    let verified_msgs = VERIFIED_BLOCK_MESSAGES.lock().unwrap();
+    let tm_ns = verified_msgs.time();
+    return (tm_ns / 1000000000) as i64;
+}
+
+#[cfg(not(feature = "light-client-validation"))]
+fn get_current_block_time_s() -> i64
+{
+    return 0 as i64;
+}
+
+
+pub fn split_combined_cert(
+    cert: *const u8,
+    cert_len: u32,
+) -> (Vec<u8>, Vec<u8>, Vec<u8>)
+{
+    let mut vec_cert : Vec<u8> = Vec::new();
+    let mut vec_quote : Vec<u8> = Vec::new();
+    let mut vec_coll : Vec<u8> = Vec::new();
+
+    let n0 = mem::size_of::<u32>() as u32 * 3;
+
+    if cert_len >= n0
+    {
+        let p_cert = cert as *const u32;
+        let s0 = u32::from_le( unsafe { *p_cert } );
+        let s1 = u32::from_le( unsafe { *(p_cert.offset(1)) } );
+        let s2 = u32::from_le( unsafe { *(p_cert.offset(2)) } );
+
+        let size_total =
+            (n0 as u64) +
+            (s0 as u64) +
+            (s1 as u64) +
+            (s2 as u64);
+
+        if size_total <= cert_len as u64
+        {
+            vec_cert = unsafe { slice::from_raw_parts(cert.offset(n0 as isize), s0 as usize).to_vec() };
+            vec_quote = unsafe { slice::from_raw_parts(cert.offset((n0 + s0) as isize), s1 as usize).to_vec() };
+            vec_coll = unsafe { slice::from_raw_parts(cert.offset((n0 + s0 + s1) as isize), s2 as usize).to_vec() };
+        }
+
+    }
+
+    (vec_cert, vec_quote, vec_coll)
+}
```

Comment:
Try to avoid using unsafe - by doing this you're losing every guarantee that Rust provides. Unless you're doing FFI or something you should be using safe functions everywhere. Also makes your code way, way more readable. Something like (I just chatgpt'd it, don't take this as gospel):

```rust
fn split_combined_cert(cert_slice: &[u8]) -> (Vec<u8>, Vec<u8>, Vec<u8>) {
    let mut vec_cert: Vec<u8> = Vec::new();
    let mut vec_quote: Vec<u8> = Vec::new();
    let mut vec_coll: Vec<u8> = Vec::new();

    // This assumes the format of the cert_slice starts with three u32 integers indicating the length of each part.
    let header_size = std::mem::size_of::<u32>() * 3;

    if cert_slice.len() >= header_size {
        // Attempt to read the sizes from the slice.
        // We're using byteorder crate for more readable handling of byte order conversions.
        // Ensure you have byteorder in your Cargo.toml!
        // byteorder = "*"
        use byteorder::{ByteOrder, LittleEndian};

        let s0 = LittleEndian::read_u32(&cert_slice[0..4]) as usize;
        let s1 = LittleEndian::read_u32(&cert_slice[4..8]) as usize;
        let s2 = LittleEndian::read_u32(&cert_slice[8..12]) as usize;

        // Verify we have enough data in the slice to cover these sizes
        if cert_slice.len() >= header_size + s0 + s1 + s2 {
            let base = header_size;
            vec_cert = cert_slice[base..base + s0].to_vec();
            vec_quote = cert_slice[base + s0..base + s0 + s1].to_vec();
            vec_coll = cert_slice[base + s0 + s1..base + s0 + s1 + s2].to_vec();
        }
    }

    (vec_cert, vec_quote, vec_coll)
}
```